README.md
# ZombieOutbreakl
dadafd


changeLog.txt
# December 27:
    Moving over to the grid paradigm for large-scale structure of the UX
    Bar Color is now supported from the resources CSV
    Added a rover window and an editable number of rovers
    Rovers now effect power consumption - first interactive sim!
    Moved LogWindow into a seperate class & file
    Moved MapWindow into a seperate class & file
    Moved RoverWindow into a seperate class & file
# December 26:
    Added Kyle and Max to the GitHub repro
    created a Tech Tree Labs Slack Team and added Kyle and Max
    cleaned up all of the warnings across all the files, except disagreed with not approving camelCase
    added the map view
# December 25:
    implemented the resource bar class
        used a window canvas to limit the size of the progress bars width
    loaded resources from a CSV file
    linked the sim loop to dynamically updating the resource bars
# December 18?, 2016
    implemented the log window class, first scrolling day by day use of the simulation




LogWindow.py
import tkinter as tk


class LogWindow:
    def __init__(self, parent):
        self.logWindow = tk.LabelFrame(parent, text="Log Window", borderwidth=2, relief="sunken")

        self.headerFrame = tk.Frame(self.logWindow)

        headers = [["Day", "Survivors", "Buildings Own", "Delta", "Defense  ", "food", "water"]]
        headerLine1 = ""
        for h in headers:
            headerLine1 = "{: ^4} {: ^8} {: ^8} {: ^8} {: ^8} {: ^8} {: ^8}".format(*h)

        # headers2 = [["24h", "kW    ", "kW    ", "kW   ", "kW          ", "Lb (Billions)     ", "Lb (Trillions)     "]]
        headers2 = [["24h", "How Many", "How many", "", "Defense points", "Lb (Billions)", "Lb (Trillions)"]]
        headerLine2 = ""
        for h in headers2:
            headerLine2 = "{: ^4} {: ^8} {: ^8} {: ^8} {: ^8} {: ^8} {: ^8}".format(*h)

        self.header1 = tk.Label(self.headerFrame, text=headerLine1, font=("Courier", 12))
        self.header2 = tk.Label(self.headerFrame, text=headerLine2, font=("Courier", 12))

        self.header1.pack()
        self.header2.pack()
        self.headerFrame.pack()

        self.logTerm = tk.Frame(self.headerFrame)
        self.logTerminalScroll = tk.Scrollbar(self.logTerm, orient=tk.VERTICAL)
        self.logTerminalScroll.pack(side=tk.RIGHT, fill=tk.Y)
        self.logTerminal = tk.Listbox(self.logTerm, width=60, height=20, yscrollcommand=self.logTerminalScroll.set,
                                      font=("Courier", 12))
        self.logTerminal.pack(fill=tk.BOTH)
        self.logTerminalScroll.config(command=self.logTerminal.yview)
        self.logTerm.pack()
        self.logWindow.pack(side=tk.LEFT, fill=tk.X, expand=1)

    def printTerminal(self, lines):
        for line in lines:
            self.logTerminal.insert(0, line)  # pops it at the top of the list box
            self.logTerminal.select_clear(self.logTerminal.size() - 2)  # Clear the current selected item
            self.logTerminal.select_set(tk.END)  # Select the new item
            self.logTerminal.yview(tk.END)


SMarsMain.py
# library imports
import sys
import tkinter as tk

# SMars imports
import SMarsResourceBar
import SMarsResource
import LogWindow
import MapWindow
import RoverWindow


class Application(tk.Frame):
    def __init__(self, master=None):
        self.mainFrame = tk.Frame.__init__(self, master)

        self.day = 0  # starts on day 0 of landing
        self.excessPower = 0
        self.totalSolarPowerInstalled = 0  # kilo watts
        self.solarPanelsInstalled = 0
        self.solarPanelsStock = 0
        self.totalPowerConsumed = 0
        self.roverShiftsPerDay = 0
        self.roverBatteryBank = 0  # kW
        self.powerDelta = 0
        self.Food = 0  # kg
        self.Water = 0  # kg
        self.solarHoursPerDay = 0
        self.solarIrradianceCurve = 0  # irradiance
        self.panelsInstalledPerShift = 0  # see time break down
        self.panelPowerRating = 0.0  # kW Each panel is 260 watts on Earth * 58% efficient on Mars
        self.maxDays = 0  #
        self.powerDeltaCopy = 0  # for record keeping checks
        self.numRovers = 0
        self.numBatteries = 0
        self.batterySize = 0.0  # kW
        self.batteryMinCharge = 0.0  # minimum % charge on the batteries as a floor
        self.maxBatteryBank = self.numBatteries * self.batterySize
        self.batteryBank = 0  # assume that we land with batteries fully charged
        self.fuelCellConsumption = 0.0  # kg per kW

        self.logWindow = None
        self.mapWindow = None

        self.roverWindow = None

        self.resourceManager = None
        self.powerFrame = None
        self.powerUsedBar = None
        self.solarPowerBar = None
        self.batteryBankBar = None
        self.resourcesFrame = None
        self.oxygenResBar = None
        self.methaneResBar = None
        self.waterResBar = None
        self.CO2ResBar = None

        self.getInitialConditions()
        self.resourceManager = SMarsResource.ResourceManager()
        self.setupUX()

    def setupUX(self):
        self.setupPowerBars(self.mainFrame)
        self.powerFrame.grid(row=0, column=0)

        self.roverWindow = RoverWindow.RoverWindow(self.mainFrame, self.numRovers)
        self.roverWindow.roverWindow.grid(row=0, column=1)

        self.mapWindow = MapWindow.MapWindow(self.mainFrame)
        self.mapWindow.mapWindow.grid(row=1, column=0)

        self.logWindow = LogWindow.LogWindow(self.mainFrame)
        self.logWindow.logWindow.grid(row=1, column=1, sticky=tk.N)

        self.setupResourceBars(self.mainFrame)
        self.resourcesFrame.grid(row=1, column=2, sticky=tk.N)

        self.master.focus()
        self.master.bind('<Return>', lambda e: self.doReturn())
        self.master.bind('<space>', lambda e: self.doSpace())

    def setupPowerBars(self, parent):
        mW = 1000.0  # 1000 kW
        # solarPower Stock
        self.powerFrame = tk.LabelFrame(parent, text="Power", borderwidth=2, relief="sunken", padx=10, pady=5)
        self.powerUsedBar = SMarsResourceBar.ResourceBar(self.powerFrame,
                                                         self.resourceManager.sResources['Survivors Left'], 120, mW,
                                                         self.totalPowerConsumed, True)
        self.solarPowerBar = SMarsResourceBar.ResourceBar(self.powerFrame,
                                                          self.resourceManager.sResources['Dead Humans'],
                                                          120, mW, self.totalSolarPowerInstalled, True)
        self.batteryBankBar = SMarsResourceBar.ResourceBar(self.powerFrame,
                                                           self.resourceManager.sResources['ZombieDead'], 120, 5 * mW,
                                                           self.batteryBank, True)

    def setupResourceBars(self, parent):
        mT = 1000.0  # 1000 kgs
        self.resourcesFrame = tk.LabelFrame(parent, text="Resources", borderwidth=2, relief="sunken", padx=10,
                                            pady=5)
        self.oxygenResBar = SMarsResourceBar.ResourceBar(self.resourcesFrame,
                                                         self.resourceManager.sResources['Food'],
                                                         100, self.Water, self.Water, False)
        self.methaneResBar = SMarsResourceBar.ResourceBar(self.resourcesFrame,
                                                          self.resourceManager.sResources['Water'],
                                                          100, self.Food, self.Food, False)
        self.waterResBar = SMarsResourceBar.ResourceBar(self.resourcesFrame, self.resourceManager.sResources['Water'],
                                                        100, mT, 1, False)
        self.CO2ResBar = SMarsResourceBar.ResourceBar(self.resourcesFrame, self.resourceManager.sResources['ZombieDead'], 100,
                                                      mT, 1, False)

    def doSpace(self):
        self.simLoop()

    def doReturn(self):
        self.simLoop()

    def getInitialConditions(self):
        self.day = 0  # starts on day 0 of landing
        self.excessPower = 10000
        self.totalSolarPowerInstalled = 0  # kilo watts
        self.solarPanelsInstalled = 0
        self.solarPanelsStock = 0
        self.totalPowerConsumed = 1
        self.roverShiftsPerDay = 3  # assumes 6 hours of working and 2 hours for recharge / software updates / misc
        self.roverBatteryBank = 1000    # kW
        self.Food = 133  # kg
        self.Water = 133  # kg
        self.solarHoursPerDay = 7
        self.solarIrradianceCurve = 0.66
        self.panelsInstalledPerShift = 12  # see time break down
        self.panelPowerRating = 0.150  # kW Each panel is 260 watts on Earth * 58% efficient on Mars
        self.maxDays = 100  # for now
        self.powerDeltaCopy = 0  # for record keeping checks
        self.numRovers = 1
        self.numBatteries = 0
        self.batterySize = 13.5


        # kW
        self.batteryMinCharge = 0.5  # minimum % charge on the batteries as a floor
        self.maxBatteryBank = self.numBatteries * self.batterySize
        self.batteryBank = self.maxBatteryBank  # assume that we land with batteries fully charged
        self.fuelCellConsumption = 0.01  # kg per kW

    def updateSim(self):
        # Check for fail condition
        if (self.Food < 0) or (self.Water < 0):
            self.doSimQuit()

        # calculate new power installed each day
        powerPerPanel = self.panelPowerRating * self.solarHoursPerDay * self.solarIrradianceCurve
        powerInstalledPerDay = self.numRovers * self.roverShiftsPerDay * self.panelsInstalledPerShift * powerPerPanel
        self.totalSolarPowerInstalled += powerInstalledPerDay

        # calculate power consumed each day
        powerConsumedEachDay = self.numRovers * self.roverBatteryBank * self.roverShiftsPerDay
        self.totalPowerConsumed = powerConsumedEachDay

        # do the power accounting
        self.powerDelta = self.totalSolarPowerInstalled - self.totalPowerConsumed
        self.powerDeltaCopy = self.powerDelta

        self.excessPower = 0

        # if powerDelta is positive, we have extra energy, if it is negative we need from storage
        if self.powerDelta >= 0:
            print("Extra Power!")
            # add the extra power to the battery bank
            self.batteryBank += self.powerDelta
            # We cannot charge over the max size of the battery bank
            if self.batteryBank > self.maxBatteryBank:
                # find out what the excess power is after charging the batteries to full
                self.powerDelta = self.batteryBank - self.maxBatteryBank
                # find out what the excess power is after charging the batteries to full
                self.batteryBank = self.maxBatteryBank  # set the current bank to max bank
                # self.powerDelta now contains excess power available for other systems
                self.excessPower = self.powerDelta

        else:
            # take power from batteries first, second from fuel cells
            batteryAvailable = self.batteryBank - self.maxBatteryBank * self.batteryMinCharge

            if batteryAvailable > abs(self.powerDelta):  # Do we have more available than we need?
                self.batteryBank += self.powerDelta  # great take it
                self.powerDelta = 0  # set the gap to 0
            else:
                self.powerDelta += batteryAvailable  # take what is available
                # now we should be at the floor of the reserves
                self.batteryBank = self.maxBatteryBank * self.batteryMinCharge

                # the remaining energy is needed from the CH4/O2 fuel cell
                fuelUsed = self.fuelCellConsumption * abs(self.powerDeltaCopy)  # determine the kg of fuel consumed

                # Burns CH4 + 2O2  = > CO2 + 2H2O
                # O2 is approx 2x as heavy as CH4, and we need to burn 2x O2 for every CH4, so 4kg of O2
                self.Food -= fuelUsed
                self.Water-= fuelUsed
                self.powerDelta = 0  # and now we are done paying for our power


                if self.Food < 0:
                    self.logWindow.printTerminal(["Ran out of Food!"])
                if self.Water < 0:
                    self.logWindow.printTerminal(["Ran out of Water!"])

    def getUserInputs(self):
        # Text Entries
        self.numRovers = int(self.roverWindow.numRoverEntryVar.get())

    def updateUX(self):
        # Power Bars
        self.solarPowerBar.updateValue(self.totalSolarPowerInstalled)
        self.powerUsedBar.updateValue(self.totalPowerConsumed)
        self.batteryBankBar.updateValue(self.batteryBank)

        # Resource Bars
        self.oxygenResBar.updateValue(self.Water)
        self.methaneResBar.updateValue(self.Food)
        self.waterResBar.updateValue(0)
        self.CO2ResBar.updateValue(0)

    def printDayStatus(self):
        status = [[self.day, self.totalPowerConsumed, self.totalSolarPowerInstalled,
                   self.powerDeltaCopy, self.batteryBank, self.Food, self.Water]]
        statusLine = []
        for s in status:
            statusLine = "{:4d} {:8.1f} {:8.1f} {:8.1f} {:8.1f} {:8.1f} {:8.1f}".format(*s)

        if len(statusLine) > 0:
            self.logWindow.printTerminal([statusLine])

    def simLoop(self):
        self.getUserInputs()
        self.updateSim()  # update the calculations

        self.updateUX()
        self.printDayStatus()  # print the daily status

        if self.day > self.maxDays:  # check to see if we are at the end of the sim
            self.doSimQuit()
        self.day += 1

    @staticmethod
    def doSimQuit():
        print("End of Simulation")
        sys.exit()


def main():
    root = tk.Tk()
    app = Application(root)
    app.master.title(gGameName)
    app.updateUX()
    app.mainloop()
    root.quit()


gGameName = "Real Mars Simulation"

if __name__ == '__main__':
    main()


RoverWindow.py

from PIL import Image, ImageTk
import tkinter as tk


class RoverWindow:
    def __init__(self, parent, initialValue):
        self.roverWindow = tk.LabelFrame(parent, text="Zombie (In hundred-millions)", borderwidth=2, relief="sunken", padx=20, pady=10)
        self.numRoverEntryVar = tk.StringVar()

        self.roverImageLabel = tk.Label(self.roverWindow, width=32, height=32)
        image = Image.open("./assets/ZombiePic.jpeg").convert("RGB")
        image = image.resize((32, 32), Image.ANTIALIAS)
        self.roverImage = ImageTk.PhotoImage(image)
        self.roverImageLabel.configure(image=self.roverImage)
        self.roverImageLabel.image = self.roverImage
        self.roverImageLabel.pack(side=tk.LEFT)

        self.numRoverEntry = tk.Entry(self.roverWindow, textvariable=self.numRoverEntryVar)
        self.numRoverEntryVar.set(str(initialValue))
        self.numRoverEntry.pack()


SMarsResource.py
# Resource

# ID
# Name
# Units
# Icon

# Assume icons are 32x32

# Need to load this data from a csv file

import csv


class Resource:
    def __init__(self, rid):
        # Look up name, Units, and Icon from CSV file to populate the resource
        self.rid = rid


class ResourceManager:
    sResources = {}

    def __init__(self):
        self.resourcePath = './assets/Resources.csv'
        with open(self.resourcePath, 'r') as csvfile:
            resourceReader = csv.reader(csvfile, delimiter=',', quotechar='|', quoting=csv.QUOTE_MINIMAL)
            rowNumber = 0
            for row in resourceReader:
                if rowNumber > 0:  # skip the header row
                    print(rowNumber)
                    print(row)
                    r = Resource(int(row[0]))
                    r.name = row[1]
                    r.units = row[2]
                    r.icon = './assets/' + row[3]
                    r.barColor = row[4]
                    self.sResources[r.name] = r
                rowNumber += 1

        print(" Check Resources")
        for r in self.sResources:
            print(r)


SMarsResourceBar.py
# Resource Bar
# Progress Bar with min, max, current values
# Length
# Text Label
# Current value
# Color (TODO: or full, medium, low)
# Image label

from PIL import Image, ImageTk
import tkinter as tk
import tkinter.ttk as ttk


class ResourceBar:
    def __init__(self, parent, resource, barLength, maxValue, initialValue, horiz):
        # rid is the resource id
        # barColor is text 'red', 'blue' etc

        name = resource.name
        units = resource.units
        fullText = name + " (" + units + ")"
        iconPath = resource.icon
        barColor = resource.barColor

        # Create a LabelFrame to hold this resource bar
        self.resourceFrame = tk.LabelFrame(parent, text=fullText, borderwidth=1, relief="sunken")

        # Create the Label to hold the icon
        self.resourceIconLabel = tk.Label(self.resourceFrame, width=32, height=32)

        # Load the icon and pop it into the Label
        image = Image.open(iconPath).convert("RGB")
        image = image.resize((38

                                   , 38), Image.ANTIALIAS)
        self.icon = ImageTk.PhotoImage(image)
        self.resourceIconLabel.configure(image=self.icon)
        self.resourceIconLabel.image = self.icon
        self.resourceIconLabel.pack(side=tk.LEFT, padx=5)

        # Setup the progress bar
        self.value = tk.DoubleVar()
        self.value.set(initialValue)
        self.valuetext = tk.StringVar()
        self.valuetext.set(str(initialValue))

        # self.pbcanvas = tk.Canvas(self.resourceFrame, relief=tk.FLAT, background="#D2D2D2", width=barLength, height=5)
        self.pbcanvas = tk.Canvas(self.resourceFrame, relief=tk.FLAT,  width=barLength, height=5)
        s = ttk.Style()
        s.theme_use('alt')
        s.configure(barColor+".resource.Horizontal.TProgressbar", foreground=barColor, background=barColor)
        print(barColor)
        self.pb = ttk.Progressbar(self.pbcanvas, variable=self.value,
                                  style=barColor+".resource.Horizontal.TProgressbar",
                                  orient="horizontal", max=maxValue, length=barLength,
                                  value=initialValue, mode="determinate")
        self.pb.pack(padx=10)

        self.pbcanvas.create_window(1, 1, anchor=tk.NW, window=self.pb)
        self.pbcanvas.pack(pady=1)

        self.resourceTextLabel = tk.Label(self.resourceFrame, textvariable=self.valuetext)
        self.resourceTextLabel.pack()

        self.resourceIconLabel.pack()

        if horiz:
            self.resourceFrame.pack(side=tk.LEFT)
        else:
            self.resourceFrame.pack()

    def updateValue(self, value):
        self.value.set(value)
        self.valuetext.set(str("{0:.1f}".format(round(value, 1))))


# NOTES
# ProgressBar Styles: classic, clam, alt, xpnative (the default on my system), and winnative Alt looks the best


SMarsMainBasicSim.py
from Tkinter import *

try:
    import Image
    import ImageDraw
    import ImageFont
    import ImageTk
except ImportError:
    from PIL import Image, ImageDraw, ImageFont, ImageTk


class Application(Frame):
    def __init__(self, master=None):
        self.mainFrame = Frame.__init__(self, master)
        self.day = 0  # starts on day 0 of landing
        self.excessPower = 0
        self.totalSolarPowerInstalled = 0  # kilo watts
        self.solarPanelsInstalled = 0
        self.solarPanelsStock = 0
        self.totalPowerConsumed = 0
        self.roverShiftsPerDay = 0
        self.roverBatteryBank = 0  # kW
        self.powerDelta = 0
        self.reservesCH4 = 0  # kg
        self.reservesO2 = 0  # kg
        self.solarHoursPerDay = 0
        self.solarIrradianceCurve = 0  # irradiance
        self.panelsInstalledPerShift = 0  # see time break down
        self.panelPowerRating = 0.0  # kW Each panel is 260 watts on Earth * 58% efficient on Mars
        self.maxDays = 0  #
        self.powerDeltaCopy = 0  # for record keeping checks
        self.numRovers = 0
        self.numBatteries = 0
        self.batterySize = 0.0  # kW
        self.batteryMinCharge = 0.0  # minimum % charge on the batteries as a floor
        self.maxBatteryBank = self.numBatteries * self.batterySize
        self.batteryBank = 0  # assume that we land with batteries fully charged
        self.fuelCellConsumption = 0.0  # kg per kW

        self.logWindow = None
        self.logTerminal = None
        self.logTerminalScroll = None
        self.header1 = None
        self.header2 = None
        self.logTerm = None
        self.headerFrame = None
        self.resourceManager = None
        self.powerFrame = None
        self.powerUsedBar = None
        self.solarPowerBar = None
        self.batteryBankBar = None
        self.resourcesFrame = None
        self.oxygenResBar = None
        self.methaneResBar = None
        self.waterResBar = None
        self.CO2ResBar = None

        self.createLogWindow(self.mainFrame)

        self.master.focus()
        self.master.bind('<Return>', lambda e: self.doReturn())
        self.master.bind('<space>', lambda e: self.doSpace())

    def doSpace(self):
        self.simLoop()

    def doReturn(self):
        self.simLoop()

    def centerWindow(self):
        w = 1000
        h = 800

        sw = self.master.winfo_screenwidth()
        sh = self.master.winfo_screenheight()

        x = (sw - w) / 2
        y = (sh - h) / 2
        self.master.geometry('%dx%d+%d+%d' % (w, h, x, y))

    def printTerminal(self, lines):
        for line in lines:
            self.logTerminal.insert('end', line)
            self.logTerminal.select_clear(self.logTerminal.size() - 2)  # Clear the current selected item
            self.logTerminal.select_set(END)  # Select the new item
            self.logTerminal.yview(END)

    def createLogWindow(self, parent):
        self.logWindow = LabelFrame(parent, text="Log Window", borderwidth=2, relief="sunken", width=100, height=100)
        self.logTerm = Frame(self.logWindow, width=800, height=100)
        self.logTerminalScroll = Scrollbar(self.logTerm, orient=VERTICAL)
        self.logTerminalScroll.pack(side=RIGHT, fill=Y)
        self.logTerminal = Listbox(self.logTerm, width=80, height=25, yscrollcommand=self.logTerminalScroll.set)
        self.logTerminal.pack(side=LEFT, fill=BOTH)
        self.logTerminalScroll.config(command=self.logTerminal.yview)
        self.logTerm.pack(side=LEFT)
        self.logWindow.pack(fill=X, expand=1)

    def getInitialConditions(self):
        self.roverShiftsPerDay = 1  # assumes 6 hours of working and 2 hours for recharge / software updates / misc
        self.roverBatteryBank = 10000

        # kW
        self.reservesCH4 = 100  # kg
        self.reservesO2 = 100  # kg
        self.solarHoursPerDay = 7
        self.solarIrradianceCurve = 0.66
        self.panelsInstalledPerShift = 12  # see time break down
        self.panelPowerRating = 0  # kW Each panel is 260 watts on Earth * 58% efficient on Mars
        self.maxDays = 1000  # for now
        self.numRovers = 2
        self.Food = 1
        self.Zombies = 1 * 100

    def updateSim(self):
        # calculate new power installed each day
        powerPerPanel = self.Food = self.powerPerPanel
        powerConsumedEachDay = self.Zombies * self.Zombies
        self.totalPowerConsumed = powerConsumedEachDay

    def printDayStatus(self):
        # Every tenth day draw the column headers
        lines = []
        if self.day % 10 == 0:
            lines.append("Day | Power- | Solar+ ")
        lines.append(str(self.day) + " | " + str(self.totalPowerConsumed) + " | " + str(self.totalSolarPowerInstalled))
        self.printTerminal(lines)

    def simLoop(self):
        self.updateSim()  # update the calculations
        self.printDayStatus()  # do the calculations
        if self.day > self.maxDays:  # check to see if we are at the end of the sim
            self.doSimQuit()
        self.day += 1

    @staticmethod
    def doSimQuit():
        print("End of Simulation")
        sys.exit()


def main():
    root = Tk()
    app = Application(root)
    app.master.title(gGameName)
    app.getInitialConditions()
    app.mainloop()
    root.quit()


gGameName = "Real Mars Simulation"

if __name__ == '__main__':
    main()
